### PR TITLE
Add Ellipsis typehints

### PR DESCRIPTION
**Pull request by headtr1ck**
This PR adds an `Ellipsis` typehint to some functions.

Interestingly mypy did not complain at the tests before, I assume it is because "..." is Hashable or something like that?

I don't know what to do with reductions, since they also support ellipsis, but it is basically the same as using None.
Therefore, I assume it is not necessary to expose this feature.

Did I miss any functions where ellipsis is supported? It is hard to look for "..."... xD

### CODE REVIEW

File: xarray/core/types.py [hunk 1]
```diff
@@ -29,13 +30,19 @@
...
+
 else:
     Self: Any = None
+    Ellipsis: Any = None
+
 
 T_Dataset = TypeVar("T_Dataset", bound="Dataset")
 T_DataArray = TypeVar("T_DataArray", bound="DataArray")
 T_Variable = TypeVar("T_Variable", bound="Variable")
+T_Array = TypeVar("T_Array", bound="AbstractArray")
```
Comment (headtr1ck) [inline xarray/core/types.py:45]: whoops that was not supposed to be checked in, but I guess it doesn't hurt.

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -3769,7 +3770,7 @@ def imag(self: T_DataArray) -> T_DataArray:
     def dot(
         self: T_DataArray,
         other: T_DataArray,
-        dims: Hashable | Sequence[Hashable] | None = None,
+        dims: str | Iterable[Hashable] | Ellipsis | None = None,
```
Comment (max-sixty) [inline xarray/core/dataarray.py:3773]: Excellent! Great that we can roll the `str | Iterable[Hashable]` out

**Review by max-sixty — APPROVED** [review]
Thanks! Another good one!

Do you know whether `Ellipsis` works now? I had put an issue in with mypy, and discussed it briefly (with Guido!), is it now in mypy? (and regardless, this will be nicely forward-compatible)

**Comment by headtr1ck** [comment]
> Do you know whether `Ellipsis` works now? I had put an issue in with mypy, and discussed it briefly (with Guido!), is it now in mypy? (and regardless, this will be nicely forward-compatible)

I saw that issue :)
That was the only way I could get it to work, in runtime it still does not seem to work. Maybe "Ellipsis" (with "") could work?

File: xarray/core/dataarray.py (hunk 2 above)
Comment (headtr1ck) [reply xarray/core/dataarray.py:3773]: Yes, as soon as you try to type some internals you will run into problems with `Hashable | Sequence[Hashable]`.
This really seems to be the only way.

We should probably add a parser for this in utils such that all methods handle it the same.
Maybe I can come up with a PR for that.

**Comment by max-sixty** [comment]
That makes sense. Until mypy make the change, I think this is a nice signpost for users, and will be easily forward compatible…

**Comment by dcherian** [comment]
> I don't know what to do with reductions, since they also support ellipsis, but it is basically the same as using None.

For Groupby reductions `...` means reduce over all dimensions; `None` means reduce over `groups`'s dimensions

**Pull request merged**